File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
int ROWS;
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        ROWS = prerequisites.size();
        vector<vector<int>> adjacent(numCourses);
        vector<int> in_arrows(numCourses, 0);

        queue<int> queue;

        for (auto& p : prerequisites) {
            int u = p[0];
            int v = p[1];
            adjacent[v].push_back(u);
            in_arrows[u]++;
        }

        for (int i = 0; i < numCourses; i++) {
            if (in_arrows[i] == 0) {
                queue.push(i);
            }
        }
        int courses_taken = 0;
        vector<int> result;
        
        while (!queue.empty()) {
            int curr = queue.front();
            queue.pop();
            result.push_back(curr);
            courses_taken++;

            for (auto& adj : adjacent[curr]) {
                in_arrows[adj]--;
                if (in_arrows[adj] == 0) {
                    queue.push(adj);
                }
            }
        
        }
    
        // if (courses_taken == numCourses) {
        //     return result
        // }

        return (courses_taken == numCourses) ? result : vector<int>{};
    }
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: list[list[int]]) -> list[int]:

        preMap = {i:[] for i in range(numCourses)}

        for pre, crs in prerequisites:
            preMap[pre].append(crs)
        
        cycle = set()
        output = []

        def dfs(crs):
            if crs in cycle:
                return False
            
            if preMap[crs] is None:
                return True
            
            cycle.add(crs)

            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            
            preMap[crs] = None
            cycle.remove(crs)
            output.append(crs)
            return True
        
        for c in range(numCourses):
            if not dfs(c):
                return []

        return output
        